pystiche_papers/sanakoyeu_et_al_2018/_utils.py

```diff
@@ -0,0 +1,107 @@
+from typing import Any, Iterator, List, Optional, Union
+
+import torch
+from torch import nn, optim
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.optimizer import Optimizer
+
+from pystiche.ops.container import OperatorContainer
+from pystiche.ops.op import ComparisonOperator
+from pystiche.optim.meter import FloatMeter
+
+
+def optimizer(
+    params: Union[nn.Module, Iterator[torch.nn.parameter.Parameter]]
+) -> Optimizer:
+    r"""Optimizer from :cite:`SKL+2018`.
+
+    Args:
+        params: Parameter to be optimized.
+
+    """
+    if isinstance(params, nn.Module):
+        params = params.parameters()
+    return optim.Adam(params, lr=2e-4)
+
+
+def lr_scheduler(optimizer: Optimizer) -> Optional[ExponentialLR]:
+    r"""Learning rate scheduler from :cite:`SKL+2018`.
+
+    Args:
+        optimizer: Wrapped optimizer.
+
+    """
+    return DelayedExponentialLR(optimizer, gamma=0.1, delay=2)
+
+
+class ExponentialMovingAverage(FloatMeter):
+    def __init__(
+        self,
+        name: str,
+        init_val: float = 0.8,
+        smoothing_factor: float = 0.05,
+        fmt: str = "{:f}",
+    ) -> None:
+        super().__init__(name)
+        self.smoothing_factor = smoothing_factor
+        self.last_val = init_val
+        self.fmt = fmt
+
+    def calculate_val(self, val: Union[torch.Tensor, float]) -> float:
+        if isinstance(val, torch.Tensor):
+            val = val.item()
+        return (
+            self.last_val * (1.0 - self.smoothing_factor) + self.smoothing_factor * val
+        )
+
+    def update(self, new_val: Union[torch.Tensor, float]) -> None:  # type: ignore[override]
+        super().update(self.calculate_val(new_val))
+
+    def local_avg(self) -> float:
+        return self.last_val
```

Comment:
How is `last_val` an average of some kind?

Reply:
The `ExponentialMovingAverageMeter` tracks the course of the value. Therefore, the `last_val` corresponds to the current average value calculated in `calculate_val` in each `update`.

Reply:
I see. But then, isn't it a `global_avg` since it includes all values ever recorded?